src/cryptonote_core/loki_name_system.cpp

```diff
@@ -0,0 +1,1124 @@
+        *reason = err_stream.str();
+      }
+      return false;
+    }
+  }
+  return true;
+}
+
+static std::stringstream &error_stream_append_tx_msg(std::stringstream &err_stream, cryptonote::transaction const &tx, cryptonote::tx_extra_loki_name_system const &data)
+{
+  err_stream << "TX type=" << tx.type << ", tx=" << get_transaction_hash(tx) << ", owner=" << data.owner << ", type=" << (int)data.type << ", name=" << data.name << ", ";
+  return err_stream;
+}
+
+static bool validate_against_previous_mapping(lns::name_system_db const &lns_db, uint64_t blockchain_height, cryptonote::transaction const &tx, cryptonote::tx_extra_loki_name_system const &data, std::string *reason = nullptr)
+{
+  std::stringstream err_stream;
+  crypto::hash expected_prev_txid = crypto::null_hash;
+  if (lns::mapping_record mapping = lns_db.get_mapping(data.type, data.name.data(), data.name.size()))
+  {
+    if (data.type != static_cast<uint16_t>(lns::mapping_type::lokinet))
+    {
+      if (reason)
+      {
+        lns::user_record user = lns_db.get_user_by_id(mapping.user_id);
+        error_stream_append_tx_msg(err_stream, tx, data) << "non-lokinet entries can NOT be renewed, mapping already exists with name=" << mapping.name << ", owner=" << user.key << ", type=" << mapping.type;
+        *reason = err_stream.str();
+      }
+      return false;
+    }
+
+    if (!mapping.active(lns_db.network_type(), blockchain_height))
+      return true;
+
+    expected_prev_txid                 = mapping.txid;
+    uint64_t renew_window              = 0;
+    uint64_t expiry_blocks             = lns::lokinet_expiry_blocks(lns_db.network_type(), &renew_window);
+    uint64_t const renew_window_offset = expiry_blocks - renew_window;
+    uint64_t const min_renew_height    = mapping.register_height + renew_window_offset;
+
+    if (min_renew_height >= blockchain_height)
+    {
+      error_stream_append_tx_msg(err_stream, tx, data) << "trying to renew too early, the earliest renew height=" << min_renew_height << ", urrent height=" << blockchain_height;
+      *reason = err_stream.str();
+      return false; // Trying to renew too early
+    }
+
+    // LNS entry can be renewed, check that the request to renew originates from the owner of this mapping
+    lns::user_record requester = lns_db.get_user_by_key(data.owner);
+    if (!requester)
+    {
+      if (reason)
+      {
+        error_stream_append_tx_msg(err_stream, tx, data) << "trying to renew existing mapping but owner specified in LNS extra does not exist, rejected";
+        *reason = err_stream.str();
+      }
+      return false;
+    }
+
+    lns::user_record owner = lns_db.get_user_by_id(mapping.user_id);
+    if (!owner)
+    {
+      if (reason)
+      {
+        error_stream_append_tx_msg(err_stream, tx, data) << "unexpected user_id=" << mapping.user_id << " does not exist";
+        *reason = err_stream.str();
+      }
+      return false;
+    }
+
+    if (requester.id != owner.id)
```

Comment:
This check is insufficient.  As currently written this allows someone to renew my .loki registration *and change the value* as long as they leave me as the owner.

But also, even if a renewal can't change the value, I don't think someone who isn't the owner should be able to renew my domain without my permission.  I realize we want to allow someone else paying for it, but we can find a better way to allow that on renewals: for instance, by having the owner provide a signed value that references the previous registration (which the third-party purchaser would include in the renewal tx).